---
title: "FADS_prioritised_variants_lolliplots"
author: "Ella Davyson"
date: "2024-11-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyr)
library(dplyr)
library(trackViewer)
library(ggplot2)
library(ggpattern)
```

```{r}
########### PRIORITISED VARIANTS IN THE FADS CLUSTER ############
# Can run this locally (just variant information)
setwd("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files")
FADS <- read.table("FADS_cluster_UKB_pVCF.tsv", sep = "\t", header = T)

# Read in the priority variant tables 

for (i in c(1:nrow(FADS))){
  gene <- FADS$hgnc_symbol[i]
  priority <- read.table(paste0("vep_res/AC_overzero/", gene, "_vepoutput_proc/", gene, "_priority_annot.tsv"), sep = "\t", header = T)
  priority_mdd <- read.table(paste0("pri_variant_files/mdd_metabol_cohortvars/", gene, "_priority_var_mdd.tsv"), sep = "\t", header =T)
  priority_metabol <- read.table(paste0("pri_variant_files/mdd_metabol_cohortvars/", gene, "_priority_var_metabol.tsv"), sep = "\t", header =T)
  pri_csq_mdd <- read.table(paste0("vep_res/AC_overzero_06_01/", gene, "_vepoutput_proc/", gene, "_priority_consequence_MDD_cohort.tsv"), sep = "\t", header = T)
  pri_csq_metabol <- read.table(paste0("vep_res/AC_overzero_06_01/", gene, "_vepoutput_proc/", gene, "_priority_consequence_Metabolite_cohort.tsv"), sep = "\t", header = T)
  pri_ACcounts_mdd <- read.table(paste0("vep_res/AC_overzero_06_01/", gene, "_vepoutput_proc/", gene, "_priorityACcounts_MDD_cohort.tsv"), sep = "\t", header = T)
   pri_ACcounts_metabol <- read.table(paste0("vep_res/AC_overzero_06_01/", gene, "_vepoutput_proc/", gene, "_priorityACcounts_Metabolite_cohort.tsv"), sep = "\t", header = T)
  print(paste0("Read in priority variants: ", gene, " (n=", nrow(priority), ")"))
  assign(paste0(gene, "_priority_var"), priority)
  assign(paste0(gene, "_priority_mdd"), priority_mdd)
  assign(paste0(gene, "_priority_metabol"), priority_metabol)
  assign(paste0(gene, "_priority_csq_mdd"), pri_csq_mdd)
  assign(paste0(gene, "_priority_csq_metabol"), pri_csq_metabol)
  assign(paste0(gene, "_priority_ACcounts_mdd"), pri_ACcounts_mdd)
  assign(paste0(gene, "_priority_ACcounts_metabol"), pri_ACcounts_metabol)
}

```

### Overlap in FADS1 and FADS2 priortised variants ? 

FADS1 and FADS2 overlap in positions as they are on opposing strands, FADS2 (Forward strand) encompasses the whole of FADS1. Does FADS2 have all the same variants as FADS1? 

Yes, they are all included
```{r}
table(FADS1_priority_var$chr_pos_ref_alt %in% FADS2_priority_var$chr_pos_ref_alt)
```

## Consequence and AC plots of the prioritised variants 

```{r}
####### Consequence plots 
pri_csq_mdd_all <- rbind(FADS1_priority_csq_mdd, FADS2_priority_csq_mdd, FADS3_priority_csq_mdd, FEN1_priority_csq_mdd, MYRF_priority_csq_mdd, TMEM258_priority_csq_mdd) %>% mutate(cohort = paste0(cohort, "-cohort"), IMPACT = factor(IMPACT, levels = c("HIGH", "MODERATE", "LOW",  "MODIFIER")))

pri_csq_metabol_all <- rbind(FADS1_priority_csq_metabol, FADS2_priority_csq_metabol, FADS3_priority_csq_metabol, FEN1_priority_csq_metabol, MYRF_priority_csq_metabol, TMEM258_priority_csq_metabol) %>% mutate(cohort = paste0(cohort, "-cohort"), IMPACT = factor(IMPACT, levels = c("HIGH", "MODERATE", "LOW",  "MODIFIER")))

pri_csq_all <- rbind(pri_csq_mdd_all, pri_csq_metabol_all)

custom_colors <- c("HIGH" = "red", "MODERATE" = "orange", "LOW" = "#90EE90", "MODIFIER" ="lightpink")


mdd_cohort_priority <- ggplot(pri_csq_mdd_all, aes(x = reorder(Consequence, csq_count), y = csq_count, fill = IMPACT)) +
  geom_bar(stat = 'identity') + 
 theme_bw() + 
labs(x = "Consequence", y = "N", title = "Prioritised variants in the MDD cohort", 
fill = "VEP IMPACT") +
coord_flip() + 
geom_text(aes(label = csq_count), hjust = -0.1) + 
 theme(legend.position = "top") + 
scale_fill_manual(values = custom_colors) + 
facet_wrap(~GENE)

metabolite_cohort_priority <- ggplot(pri_csq_metabol_all, aes(x = reorder(Consequence, csq_count), y = csq_count, fill = IMPACT)) +
  geom_bar(stat = 'identity') + 
 theme_bw() + 
labs(x = "Consequence", y = "N", title = "Prioritised variants in the MDD cohort", 
fill = "VEP IMPACT") +
coord_flip() + 
geom_text(aes(label = csq_count), hjust = -0.1) + 
 theme(legend.position = "top") + 
scale_fill_manual(values = custom_colors) + 
facet_wrap(~GENE)

all_cohort_pri_csq <- ggplot(pri_csq_all %>% filter(csq_count > 1), aes(x = reorder(Consequence, csq_count), y = csq_count, fill = IMPACT)) +
  geom_bar(stat = 'identity') + 
 theme_bw() + 
labs(x = "Consequence", y = "N", title = "", 
fill = "VEP IMPACT") +
coord_flip() + 
geom_text(aes(label = csq_count), hjust = -0.1, size = 3) + 
 theme(legend.position = "top",
       strip.text = element_text(size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
      axis.text.x = element_text(angle = 45, hjust = 1) ) + 
scale_fill_manual(values = custom_colors) + 
facet_grid(cohort ~ GENE)

####### AC plots

pri_ACcounts_all <- rbind(FADS1_priority_ACcounts_mdd, FADS2_priority_ACcounts_mdd, FADS3_priority_ACcounts_mdd, FEN1_priority_ACcounts_mdd, MYRF_priority_ACcounts_mdd, TMEM258_priority_ACcounts_mdd, FADS1_priority_ACcounts_metabol, FADS2_priority_ACcounts_metabol, FADS3_priority_ACcounts_metabol, FEN1_priority_ACcounts_metabol, MYRF_priority_ACcounts_metabol, TMEM258_priority_ACcounts_metabol)
pri_ACcounts_all <- pri_ACcounts_all %>% mutate(range = factor(range, levels = c("1", "< 5", "< 10", "< 50", "< 100", "< 1000", "> 1000")), cohort = paste0(cohort, "-cohort"))

ACcounts_graph <- ggplot(pri_ACcounts_all, aes(x=range, y = n)) + 
  geom_bar(stat="identity", fill = "skyblue")+
  geom_text(aes(label = n), vjust = -0.5, color = "black", size = 3) + 
  labs(x = "UKB Allele Count", y = "Count", title="") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        strip.text.x = element_text(size = 12),
        strip.text.y = element_text(size = 12),
        plot.title = element_text(hjust = 0.5, face = "bold", size = 14)) + 
  facet_grid(cohort ~ GENE)

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_pri_variants_csq_allgenes_over1.png", all_cohort_pri_csq, width = 12, height = 10, device = "png", dpi= 300)

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_pri_variants_ACcounts_allgenes.png", ACcounts_graph, width = 12, height = 6, device = "png", dpi= 300)

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_pri_variants_subplots.png", ggarrange(all_cohort_pri_csq, ACcounts_graph, ncol = 1, nrow = 2), width = 8.3, height = 10, device = "png", dpi= 300)


```



## Generating the annotations file for REGENIE masks

The annotation file is in the following format

chrposrefalt    GENE    annotation (most deleterious)
Annotations in the order of deleteriousness: 

LoFTEE- High Confidence Loss of Function
VEP IMPACT - HIGH
HIGH REVEL score 
CADD score 
VEP IMPACT - MODERATE

Mask 1 - LoFTEE
Mask 2 - LoFTEE, VEP IMPACT HIGH
Mask 3 - LoFTEE, VEP IMPACT HIGH, HIGH REVEL SCORE (> 0.5)
Mask 4 - LoFTEE, VEP IMPACT HIGH, HIGH REVEL SCORE (> 0.5), HIGH CADD SCORE (> 20)
Mask 5 - LoFTEE, VEP IMPACT HIGH, HIGH REVEL SCORE (> 0.5), HIGH CADD SCORE (> 20), MODERATE VEP IMPACT 


```{r}
gene_symbols <- c(
  "ENSG00000149485"="FADS1",
  "ENSG00000134824"="FADS2",
  "ENSG00000221968"="FADS3",
  "ENSG00000168496"="FEN1",
  "ENSG00000124920"="MYRF",
  "ENSG00000134825"="TMEM258"
)

assign_masks <- function(df, gene) {
  df <- df %>%
    mutate(most_deleterious = 
                        case_when(LoF == "HC" | LoF == "OS" ~ "LoF",
                                  (LoF != "HC" | LoF != "OS") & IMPACT == "HIGH" ~ "H_IMPACT",
                                  (LoF != "HC" | LoF != "OS") & IMPACT != "HIGH" & REVEL >= 0.5 ~ "REVEL",
                                  (LoF != "HC" | LoF != "OS") & IMPACT != "HIGH" & (REVEL < 0.5 | is.na(REVEL)) & CADD_PHRED >= 20 ~ "CADD",
                                  (LoF != "HC" | LoF != "OS") & IMPACT != "HIGH" & (CADD_PHRED < 20 | is.na(CADD_PHRED)) & (REVEL < 0.5 | is.na(REVEL)) & IMPACT == "MODERATE" ~ "MOD_IMPACT")) %>% 
    mutate(mask1 = ifelse(most_deleterious == "LoF", 1, 0),
           mask2 = ifelse(most_deleterious == "LoF" | most_deleterious == "H_IMPACT", 1, 0),
           mask3 = ifelse(most_deleterious == "LoF" | most_deleterious == "H_IMPACT" | most_deleterious == "REVEL", 1, 0),
           mask4 = ifelse(most_deleterious == "LoF" | most_deleterious == "H_IMPACT" | most_deleterious == "REVEL" | most_deleterious == "CADD", 1, 0),
           mask5 = ifelse(most_deleterious == "LoF" | most_deleterious == "H_IMPACT" | most_deleterious == "CADD" | most_deleterious == "REVEL" | most_deleterious == "MOD_IMPACT", 1, 0))
  return(df)
}

priority_var_list <- list(FADS1_priority_var, FADS2_priority_var, FADS3_priority_var, FEN1_priority_var, MYRF_priority_var, TMEM258_priority_var)
priority_var_list <- lapply(priority_var_list, assign_masks)
priority_var_all <- do.call(rbind, priority_var_list)
mask_long <- priority_var_all %>% select(c(Gene, starts_with("mask"))) %>% 
  pivot_longer(cols = starts_with("mask"), names_to = "Mask", values_to = "Value") %>% 
  mutate(GENE = gene_symbols[Gene] %>% as.character()) %>% 
  mutate(mask_label = case_when(
    Mask == "mask1" ~ "LoF",
    Mask == "mask2" ~ "+HIGH IMPACT",
    Mask == "mask3" ~ "+REVEL > 0.5",
    Mask == "mask4" ~ "+CADD > 20",
    Mask == "mask5" ~ "+MODERATE IMPACT"
  ))

priority_var_mdd_list <- list(FADS1_priority_mdd, FADS2_priority_mdd, FADS3_priority_mdd, FEN1_priority_mdd, MYRF_priority_mdd, TMEM258_priority_mdd)
priority_var_mdd_list <- lapply(priority_var_mdd_list, assign_masks)
priority_var_mdd_all <- do.call(rbind, priority_var_mdd_list)
mdd_mask_long <- priority_var_mdd_all %>% select(c(Gene, starts_with("mask"))) %>% 
  pivot_longer(cols = starts_with("mask"), names_to = "Mask", values_to = "Value") %>% 
  mutate(GENE = gene_symbols[Gene] %>% as.character()) %>% 
  mutate(mask_label = case_when(
    Mask == "mask1" ~ "LoF",
    Mask == "mask2" ~ "+HIGH IMPACT",
    Mask == "mask3" ~ "+REVEL > 0.5",
    Mask == "mask4" ~ "+CADD > 20",
    Mask == "mask5" ~ "+MODERATE IMPACT"
  ))

priority_var_metabol_list <- list(FADS1_priority_metabol, FADS2_priority_metabol, FADS3_priority_metabol, FEN1_priority_metabol, MYRF_priority_metabol, TMEM258_priority_metabol)
priority_var_metabol_list <- lapply(priority_var_metabol_list, assign_masks)
priority_var_metabol_all <- do.call(rbind, priority_var_metabol_list)
metabol_mask_long <- priority_var_metabol_all %>% select(c(Gene, starts_with("mask"))) %>% 
  pivot_longer(cols = starts_with("mask"), names_to = "Mask", values_to = "Value") %>% 
  mutate(GENE = gene_symbols[Gene] %>% as.character()) %>% 
  mutate(mask_label = case_when(
    Mask == "mask1" ~ "LoF",
    Mask == "mask2" ~ "+HIGH IMPACT",
    Mask == "mask3" ~ "+REVEL > 0.5",
    Mask == "mask4" ~ "+CADD > 20",
    Mask == "mask5" ~ "+MODERATE IMPACT"
  ))

metabol_mask_long <- metabol_mask_long %>% mutate(mask_label=factor(mask_label, levels = c("LoF", "+HIGH IMPACT", "+REVEL > 0.5", "+CADD > 20", "+MODERATE IMPACT")))
mdd_mask_long <- mdd_mask_long %>% mutate(mask_label=factor(mask_label, levels = c("LoF", "+HIGH IMPACT", "+REVEL > 0.5", "+CADD > 20", "+MODERATE IMPACT")))

mask_numbers <- ggplot(mask_long %>% filter(Value == 1), aes(x = mask_label, fill = mask_label)) + 
  geom_bar(stat="count") + 
  facet_wrap(~GENE) + 
  theme_bw() + 
  scale_fill_brewer(palette ="Set1") + scale_color_brewer(palette="Set1")+ 
  labs(x = "Mask", y = "Count", fill = "Mask Criteria") + 
  theme(
    axis.text.x = element_blank(),  # Remove x-axis tick labels
    axis.ticks.x = element_blank()
  ) + geom_text(stat="count", aes(label = after_stat(count), color = mask_label), vjust = -0.5, size = 3, show.legend = FALSE) +
guides(color= guide_legend(label = FALSE)) + facet_wrap(~GENE)

mdd_mask_numbers <- ggplot(mdd_mask_long %>% filter(Value == 1), aes(x = mask_label, fill = mask_label)) + 
  geom_bar(stat="count") + 
  facet_wrap(~GENE) + 
  theme_bw() + 
  scale_fill_brewer(palette ="Set1") + scale_color_brewer(palette="Set1")+ 
  labs(x = "Mask", y = "Count", fill = "Mask Criteria", title = "MDD-cohort") + 
  theme(
    axis.text.x = element_blank(),  # Remove x-axis tick labels
    axis.ticks.x = element_blank()
  ) + geom_text(stat="count", aes(label = after_stat(count), color = mask_label), vjust = -0.5, size = 3, show.legend = FALSE) +
guides(color= guide_legend(label = FALSE)) + facet_wrap(~GENE)

metabol_mask_numbers <- ggplot(metabol_mask_long %>% filter(Value == 1), aes(x = mask_label, fill = mask_label)) + 
  geom_bar(stat="count") + 
  facet_wrap(~GENE) + 
  theme_bw() + 
  scale_fill_brewer(palette ="Set1") + scale_color_brewer(palette="Set1")+ 
  labs(x = "Mask", y = "Count", fill = "Mask Criteria", title = "Metabolite-cohort") + 
  theme(
    axis.text.x = element_blank(),  # Remove x-axis tick labels
    axis.ticks.x = element_blank()
  ) + geom_text(stat="count", aes(label = after_stat(count), color = mask_label), vjust = -0.5, size = 3, show.legend = FALSE) +
guides(color= guide_legend(label = FALSE)) + facet_wrap(~GENE)


ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_variant_masks_counts.png", mask_numbers, width = 10, height = 6, device = "png", dpi= 300)
ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_variant_masks_counts_mddcohort.png", mdd_mask_numbers, width = 10, height = 6, device = "png", dpi= 300)
ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_variant_masks_counts_metabolcohort.png", metabol_mask_numbers, width = 10, height = 6, device = "png", dpi= 300)
mask_numbers
```

You can also divide masks by the Allele Frequency. There is always a singletons only mask, but can set other thresholds. For each mask, we can subdivide to all (< 0.01), those (< 0.0001) and singletons. 

```{r}

add_AF_masks <- function(dataframe) {
  dataframe <- dataframe %>% mutate(
    mask1_0.01 = mask1,
    mask1_0.001 = ifelse(mask1 == 1 & AF < 0.001, 1, 0),
    mask1_0.0001 = ifelse(mask1 == 1 & AF < 0.0001, 1, 0),
    mask1_singleton = ifelse(mask1 == 1 & AC == 1, 1,0),
    mask2_0.01 = mask2,
        mask2_0.001 = ifelse(mask2 == 1 & AF < 0.001, 1, 0),
    mask2_0.0001 = ifelse(mask2 == 1 & AF < 0.0001, 1, 0),
    mask2_singleton = ifelse(mask2 == 1 & AC == 1, 1,0),
     mask3_0.01 = mask3,
        mask3_0.001 = ifelse(mask3 == 1 & AF < 0.001, 1, 0),
    mask3_0.0001 = ifelse(mask3 == 1 & AF < 0.0001, 1, 0),
    mask3_singleton = ifelse(mask3 == 1 & AC == 1, 1,0),
     mask4_0.01 = mask4,
        mask4_0.001 = ifelse(mask4 == 1 & AF < 0.001, 1, 0),
    mask4_0.0001 = ifelse(mask4 == 1 & AF < 0.0001, 1, 0),
    mask4_singleton = ifelse(mask4 == 1 & AC == 1, 1,0),
     mask5_0.01 = mask5,
        mask5_0.001 = ifelse(mask5 == 1 & AF < 0.001, 1, 0),
    mask5_0.0001 = ifelse(mask5 == 1 & AF < 0.0001, 1, 0),
    mask5_singleton = ifelse(mask5 == 1 & AC == 1, 1,0)
  ) %>% 
  select(-c(mask1, mask2, mask3, mask4, mask5))
return(dataframe)
}

priority_var_all <- add_AF_masks(priority_var_all)
priority_var_mdd_all <- add_AF_masks(priority_var_mdd_all)
priority_var_metabol_all <- add_AF_masks(priority_var_metabol_all)

mask_longformat <- function(dataframe) {
  dataframe <- dataframe %>% select(c(Gene, starts_with("mask"))) %>%
  pivot_longer(
    cols = starts_with("mask"),
    names_to = c("Mask", "AF_Threshold"),
    names_sep = "_",
    values_to="Value"
  ) %>% 
  mutate(GENE = gene_symbols[Gene] %>% as.character()) %>% 
  mutate(mask_label = case_when(
    Mask == "mask1" ~ "LoF",
    Mask == "mask2" ~ "+HIGH IMPACT",
    Mask == "mask3" ~ "+REVEL > 0.5",
    Mask == "mask4" ~ "+CADD > 20",
    Mask == "mask5" ~ "+MODERATE IMPACT"
  ))
  return(dataframe)
}

mask_af_long <- mask_longformat(priority_var_all)
mdd_mask_af_long <- mask_longformat(priority_var_mdd_all)
metabol_mask_af_long <- mask_longformat(priority_var_metabol_all)

summarise_mask_counts <- function(dataframe, cohortlab) {
  dataframe <- dataframe %>% filter(AF_Threshold == "singleton" | AF_Threshold == "0.01") %>% group_by(GENE, AF_Threshold, Mask) %>% 
    filter(Value ==1) %>% 
    summarise(count=n()) %>% mutate(mask_label = case_when(
    Mask == "mask1" ~ "LoF",
    Mask == "mask2" ~ "+HIGH IMPACT",
    Mask == "mask3" ~ "+REVEL > 0.5",
    Mask == "mask4" ~ "+CADD > 20",
    Mask == "mask5" ~ "+MODERATE IMPACT"
  ),
  cohort = paste0(cohortlab, "-cohort"),
  AF_Threshold = factor(AF_Threshold, levels = c("singleton", "0.01"))
    ) %>% 
    mutate(mask_label_factor = factor(mask_label,
                                         levels = c("LoF", "+HIGH IMPACT", "+REVEL > 0.5", "+CADD > 20", "+MODERATE IMPACT")))
  return(dataframe)
}

mdd_maskcounts <- summarise_mask_counts(mdd_mask_af_long, "MDD")
metabolite_maskcounts <- summarise_mask_counts(metabol_mask_af_long, "Metabolite")
maskcounts_all <- rbind(mdd_maskcounts, metabolite_maskcounts)

masks_bothcohorts <- ggplot(maskcounts_all, aes(x = AF_Threshold, y= count, fill = mask_label_factor)) +
  geom_bar(stat = "identity", position = position_dodge(width=0.8)) +
 geom_text(stat="identity", aes(label = count, color = mask_label_factor), vjust = -0.5, size = 2.5, show.legend = FALSE, position = position_dodge(width = 0.8)) +
guides(color= guide_legend(label = FALSE)) +
  facet_grid(cohort~GENE) +  # Facet by GENE
  scale_fill_brewer(palette = "Set2") +
  scale_color_brewer(palette="Set2")+
  labs(x = "AF Threshold", y = "Count", title = "", fill = "Mask Criteria") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        strip.text.x = element_text(size = 12),
        strip.text.y = element_text(size = 12),
         plot.title = element_text(hjust = 0.5, face = "bold", size = 14), legend.position = "top")

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_variant_masks_af_counts.png", 
masks_bothcohorts, width = 9, height = 6, device = "png", dpi= 300)
masks_bothcohorts

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_pri_variants__masks_subplots.png", ggarrange(ACcounts_graph, masks_bothcohorts, ncol = 1, nrow = 2, labels = c("A", "B")), width = 8.27, height = 11.69, device = "png", dpi= 300)

ggsave(filename= "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/FADS_pri_variants_csq_AC_masks_subplots.png", ggarrange(all_cohort_pri_csq, ACcounts_graph, masks_bothcohorts, ncol = 1, nrow = 3, labels = c("A", "B", "C"), heights = c(0.4, 0.3, 0.3)), width = 8.27, height = 11.69, device = "png", dpi= 300)


ggarrange(all_cohort_pri_csq, masks_bothcohorts)

```

The biggest change in number of variants is between singleton variants and 0.01, so just include two aaf bins in REGENIE, using --aaf-bins 0.01 
(singletons are always included as a default in addition to this flag)

Results tables : 

```{r}

```

## Lolliplots of each Mask 


```{r}
mane_transcripts <- c("FADS1"="ENST00000350997",
                      "FADS2"="ENST00000278840",
                      "FADS3"="ENST00000278829",
                      "FEN1"="ENST00000305885",
                      "MYRF"="ENST00000278836",
                      "TMEM258"="ENST00000537328")

## Saving genomic region variant objects of the genes

lolliplots <- function(cohort){
for (i in c(1:nrow(FADS))) {
  gene <- FADS$hgnc_symbol[i]
  transcript <- mane_transcripts[[gene]] %>% as.character()
Ensembl_file <- paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/gene_ensembl_files/ExonsSpreadsheet-Homo_sapiens_Transcript_Exons_", transcript, "_", gene, ".csv")
  gene_start <- FADS$start_position[i]
  gene_end <- FADS$end_position[i]
  strand <- FADS$strand[i]
  ensembl <- read_csv(Ensembl_file) %>% 
as.data.frame() %>% 
select(-Sequence) %>% 
filter(!grepl('sequence', `Exon / Intron`)) %>%
mutate(ExonIntron = ifelse(grepl('Intron',`Exon / Intron`), "Intron", "Exon"))%>% 
mutate(color_for_lolliplot = ifelse(ExonIntron=="Intron","#26913D","#07C5F5"))
  if (strand == -1) {
features <- GRanges("chr11", IRanges(start=c(ensembl$End %>% rev()),
end=c(ensembl$Start %>% rev()),
names = c(ensembl$ExonIntron %>% rev()),
))
features$fill <- c(ensembl$color_for_lolliplot %>% rev())
} else if (strand == 1) {
features <- GRanges("chr11", IRanges(start=c(ensembl$Start),
end=c(ensembl$End),
names = c(ensembl$ExonIntron),
))
features$fill <-  c(ensembl$color_for_lolliplot)
} else {
    stop("Provide strand info for plotting")
}
  # Read in prioritised variants 
    variant <- read.table(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/vep_res/AC_overzero/", gene, "_vepoutput_proc/", gene, "_priority_annot.tsv"), sep = "\t", header = T)
  # Read in variant zygosity
    variant_zygosity <- fread(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/variant_althet_counts/", gene, "_variant_zygosity.tsv"), sep = "\t")
    # Merge together 
variant <- merge(variant, variant_zygosity %>% select(-c(CHR, POS, REF, ALT)), by = "chr_pos_ref_alt")
variant <- variant %>% mutate(color_for_lolliplot_impact = case_when(IMPACT == "HIGH" ~ "red", IMPACT == "MODERATE" ~ "orange", IMPACT == "LOW" ~ "#90EE90", IMPACT == "MODIFIER" ~ "lightpink"))
priority_gene <- priority_var_all %>% filter(SYMBOL == gene)
# For each mask
if (gene != "FEN1") {
for (j in 1:5) {
  print(j)
  print(paste0("Mask", j, ": plotting 0.01 and singleton masks"))
  print("0.01")
  variant_0.01_mask <- priority_gene[priority_gene[, paste0("mask", j, "_0.01")] == 1, ]
  variant_0.01 <- variant %>% filter(chr_pos_ref_alt %in% variant_0.01_mask$chr_pos_ref_alt)
  print("Making the 0.01 variant object")
  SNP_0.01 <- as.numeric(variant_0.01$POS)
  print("Chr")
sample.gr.0.01 <- GRanges(unique(variant_0.01$CHR), 
IRanges(SNP_0.01, width=1, names=c(1:length(SNP_0.01))), 
strand=ifelse(strand == 1, "+", "-"))
print("Setting parameters")

sample.gr.0.01$color <- variant_0.01$color_for_lolliplot_impact
print("Label")
sample.gr.0.01$label.parameter.label <- ""
print("Score")
sample.gr.0.01$score <- as.numeric(variant_0.01$althet_carriers)
print("node label")
sample.gr.0.01$node.label <- as.character(variant_0.01$althet_carriers)
print("legend")
legend <- list(labels=c("HIGH" , "MODERATE", "LOW", "MODIFIER"), 
               col = c("red","orange", "#90EE90", "lightpink"), 
               fill = c("red","orange", "#90EE90", "lightpink"))
if(max(variant_0.01$althet_carriers) > 1000) {
text_gpar <- gpar(fontsize=7, col="black")
sample.gr.0.01$node.label.cex <- 0.6
} else if (max(variant_0.01$althet_carriers > 100)) {
  text_gpar <- gpar(fontsize=8, col="black")
sample.gr.0.01$node.label.cex <- 0.75
} else {
  text_gpar <- gpar(col="black")
sample.gr.0.01$node.label.cex <- 1
}
# Saving the 0.01 Mask
print(paste0("Saving the lolliplot of Mask", j, ".0.01 for", gene, "to ", gene, "_Mask", j, ".0.01_lolliplot_variants_pri.png"))
png(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/", gene, "_Mask", j, ".0.01_lolliplot_variants_pri.png"),
    width = 4000, height = 1200, res = 300, type = "cairo")
lolliplot(sample.gr.0.01, features, legend = legend, ylab = "Alt-heterozygous carriers in UKB Sample (N)", yaxis.gp = text_gpar, label.parameter.label = sample.gr.0.01$node.label, lollipop_style_switch_limit = 0)
grid.text(paste0(gene, ":Mask", j, ".AF < 0.01"), x=.5, y=.8, just="top", 
          gp=gpar(cex=1.5, fontface="bold"))
dev.off()
# Saving the singleton Mask
print("singletons")
  variant_singleton_mask <- priority_gene[priority_gene[, paste0("mask",j, "_singleton")]==1,]
  variant_singleton <- variant %>% filter(chr_pos_ref_alt %in% variant_singleton_mask$chr_pos_ref_alt)
    SNP_singleton <- as.numeric(variant_singleton$POS)
sample.gr.singleton <- GRanges(unique(variant_singleton$CHR), 
IRanges(SNP_singleton, width=1, names=c(1:length(SNP_singleton))), 
strand=ifelse(strand == 1, "+", "-"))
sample.gr.singleton$color <- variant_singleton$color_for_lolliplot_impact
sample.gr.singleton$label.parameter.label <- ""
sample.gr.singleton$score <- variant_singleton$althet_carriers
sample.gr.singleton$node.label <- as.character(sample.gr.singleton$score)
legend <- list(labels=c("HIGH" , "MODERATE", "LOW", "MODIFIER"), 
               col = c("red","orange", "#90EE90", "lightpink"), 
               fill = c("red","orange", "#90EE90", "lightpink"))
if(max(variant_singleton$althet_carriers) > 1000) {
text_gpar <- gpar(fontsize=7, col="black")
sample.gr.singleton$node.label.cex <- 0.6
} else if (max(variant_singleton$althet_carriers > 100)) {
  text_gpar <- gpar(fontsize=8, col="black")
sample.gr.singleton$node.label.cex <- 0.75
} else {
  text_gpar <- gpar(col="black")
sample.gr.singleton$node.label.cex <- 1
}
print(paste0("Saving the lolliplot of Mask", j, ".singleton for", gene, "to ", gene, "_Mask", j, ".singleton_lolliplot_variants_pri.png"))
png(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/", gene, "_Mask", j, ".singleton_lolliplot_variants_pri.png"),
    width = 4000, height = 1200, res = 300, type = "cairo")
lolliplot(sample.gr.singleton, features, legend = legend, ylab = "Alt-heterozygous carriers in UKB Sample (N)", yaxis.gp = text_gpar)
grid.text(paste0(gene, ":Mask", j, ".AF < 0.01"), x=.5, y=.8, just="top", 
          gp=gpar(cex=1.5, fontface="bold"))
dev.off()
}
} else if (gene == "FEN1") {
  for (j in 2:5) {
  print(j)
  print(paste0("Mask", j, ": plotting 0.01 and singleton masks"))
  print("0.01")
  variant_0.01_mask <- priority_gene[priority_gene[, paste0("mask", j, "_0.01")] == 1, ]
  variant_0.01 <- variant %>% filter(chr_pos_ref_alt %in% variant_0.01_mask$chr_pos_ref_alt)
  print("Making the 0.01 variant object")
  SNP_0.01 <- as.numeric(variant_0.01$POS)
  print("Chr")
sample.gr.0.01 <- GRanges(unique(variant_0.01$CHR), 
IRanges(SNP_0.01, width=1, names=c(1:length(SNP_0.01))), 
strand=ifelse(strand == 1, "+", "-"))
print("Setting parameters")

sample.gr.0.01$color <- variant_0.01$color_for_lolliplot_impact
print("Label")
sample.gr.0.01$label.parameter.label <- ""
print("Score")
sample.gr.0.01$score <- variant_0.01$althet_carriers
print("node label")
sample.gr.0.01$node.label <- as.character(sample.gr.0.01$score)
print("legend")
legend <- list(labels=c("HIGH" , "MODERATE", "LOW", "MODIFIER"), 
               col = c("red","orange", "#90EE90", "lightpink"), 
               fill = c("red","orange", "#90EE90", "lightpink"))
if(max(variant_0.01$althet_carriers) > 1000) {
text_gpar <- gpar(fontsize=7, col="black")
sample.gr.0.01$node.label.cex <- 0.6
} else if (max(variant_0.01$althet_carriers > 100)) {
  text_gpar <- gpar(fontsize=8, col="black")
sample.gr.0.01$node.label.cex <- 0.75
} else {
  text_gpar <- gpar(col="black")
sample.gr.0.01$node.label.cex <- 1
}
# Saving the 0.01 Mask
print(paste0("Saving the lolliplot of Mask", j, ".0.01 for", gene, "to ", gene, "_Mask", j, ".0.01_lolliplot_variants_pri.png"))
png(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/", gene, "_Mask", j, ".0.01_lolliplot_variants_pri.png"),
    width = 4000, height = 1200, res = 300, type = "cairo")
lolliplot(sample.gr.0.01, features, legend = legend, ylab = "Alt-heterozygous carriers in UKB Sample (N)", yaxis.gp = text_gpar)
grid.text(paste0(gene, ":Mask", j, ".AF < 0.01"), x=.5, y=.8, just="top", 
          gp=gpar(cex=1.5, fontface="bold"))
dev.off()
# Saving the singleton Mask
print("singletons")
  variant_singleton_mask <- priority_gene[priority_gene[, paste0("mask",j, "_singleton")]==1,]
  variant_singleton <- variant %>% filter(chr_pos_ref_alt %in% variant_singleton_mask$chr_pos_ref_alt)
    SNP_singleton <- as.numeric(variant_singleton$POS)
sample.gr.singleton <- GRanges(unique(variant_singleton$CHR), 
IRanges(SNP_singleton, width=1, names=c(1:length(SNP_singleton))), 
strand=ifelse(strand == 1, "+", "-"))
sample.gr.singleton$color <- variant_singleton$color_for_lolliplot_impact
sample.gr.singleton$label.parameter.label <- ""
sample.gr.singleton$score <- variant_singleton$althet_carriers
sample.gr.singleton$node.label <- as.character(sample.gr.singleton$score)
legend <- list(labels=c("HIGH" , "MODERATE", "LOW", "MODIFIER"), 
               col = c("red","orange", "#90EE90", "lightpink"), 
               fill = c("red","orange", "#90EE90", "lightpink"))
if(max(variant_singleton$althet_carriers) > 1000) {
text_gpar <- gpar(fontsize=7, col="black")
sample.gr.singleton$node.label.cex <- 0.6
} else if (max(variant_singleton$althet_carriers > 100)) {
  text_gpar <- gpar(fontsize=8, col="black")
sample.gr.singleton$node.label.cex <- 0.75
} else {
  text_gpar <- gpar(col="black")
sample.gr.singleton$node.label.cex <- 1
}
print(paste0("Saving the lolliplot of Mask", j, ".singleton for", gene, "to ", gene, "_Mask", j, ".singleton_lolliplot_variants_pri.png"))
png(paste0("/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Plots/", gene, "_Mask", j, ".singleton_lolliplot_variants_pri.png"),
    width = 4000, height = 1200, res = 300, type = "cairo")
lolliplot(sample.gr.singleton, features, legend = legend, ylab = "Alt-heterozygous carriers in UKB Sample (N)", yaxis.gp = text_gpar)
grid.text(paste0(gene, ":Mask", j, ".AF < 0.01"), x=.5, y=.8, just="top", 
          gp=gpar(cex=1.5, fontface="bold"))
dev.off()
}
}
}

```

## REGENIE Input

Saving an annotation file for REGENIE Input
chr:pos:ref:alt   Gene    Annotation
```{r}
# From each gene priority variants
anno_file <- priority_var_all %>% 
  mutate(chrposrefalt = gsub("_", ":", chr_pos_ref_alt)) %>%  
  mutate(chrposrefalt=gsub("chr", "", chrposrefalt)) %>%
  select(c(chrposrefalt, Gene, most_deleterious)) %>% 
  mutate(Gene = gene_symbols[Gene])
  write.table(anno_file, "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/REGENIE input/annotations_FADS.tsv", sep = "\t", row.names =F , quote = F, col.names = F)
  
# The REVEL > 0.5 annotations are captured in all the variants by the other annotation criteria
# Therefore do not need a whole separate mask for REVEL additions (so 4 masks overall)
  
  mask_file <- c("Mask1 LoF", "Mask2 LoF,H_IMPACT", "Mask3 LoF,H_IMPACT,REVEL", "Mask4 LoF,H_IMPACT,REVEL,CADD", "Mask5 LoF,H_IMPACT,REVEL,CADD,MOD_IMPACT")
readr::write_lines(mask_file, "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/REGENIE input/masks_FADS.txt")
write.table(mask_file, "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/REGENIE input/masks_FADS.txt", row.names=F, col.names = F, quote = F)

# Set list file 

setlist <- priority_var_all %>% 
  mutate(chrposrefalt=gsub("_", ":", chr_pos_ref_alt)) %>% 
  mutate(chrposrefalt=gsub("chr", "", chrposrefalt)) %>%
  group_by(SYMBOL) %>% summarise(variants = paste0(chrposrefalt, collapse = ",")) %>% mutate(CHR = 11)

setlist <- merge(setlist, FADS %>% select(hgnc_symbol, start_position), by.x = "SYMBOL", by.y = "hgnc_symbol")

setlist <- setlist %>% select(SYMBOL,CHR, start_position, variants)

write.table(setlist, "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/REGENIE input/setlist_FADS.tsv", sep = "\t", row.names =F , col.names = F, quote = F)

# AAF file 

aaf_file <- priority_var_all %>% 
  mutate(chrposrefalt=gsub("_", ":", chr_pos_ref_alt)) %>%   mutate(chrposrefalt=gsub("chr", "", chrposrefalt)) %>% select(chrposrefalt, AF, AC) %>% mutate(singleton=ifelse(AC==1,1,0)) %>% 
  select(-AC)

# Use the --set-singletons flag as this AAF file has provided information on whether it's a singleton

write.table(aaf_file, "/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/Intermediate_data_files/REGENIE input/aaf_FADS.tsv", sep = "\t", col.names = F, row.names = F, quote = F)
```

## Supplementary Tables for Thesis

```{r}
# load the excel file for adding sheets of other tables 
supp_wb <- loadWorkbook('/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/WGS_supp_data.xlsx')

sheet_names <- getSheetNames('/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/WGS_supp_data.xlsx')

# if statements to account for overwriting (if I want to add more columns etc I don't want it to be a new sheet, just to overwrite the old one)

###########################################################

# FADS1 priority variants 
if (!("FADS1_priority" %in% sheet_names)) {
  addWorksheet(supp_wb, "FADS1_priority")
}
writeData(supp_wb,"FADS1_priority", FADS1_priority_mdd %>% 
            mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% FADS1_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
            select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# FADS2 priority variants 
if (!("FADS2_priority" %in% sheet_names)) {
  addWorksheet(supp_wb, "FADS2_priority")
}
writeData(supp_wb,"FADS2_priority", FADS2_priority_mdd %>% 
            mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% FADS2_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
                   select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# FADS3 priority variants 

if (!("FADS3_priority" %in% sheet_names)) {
  addWorksheet(supp_wb, "FADS3_priority")
}
writeData(supp_wb,"FADS3_priority", FADS3_priority_mdd %>% 
             mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% FADS3_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
                   select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# FEN1 priority variants 
if (!("FEN1_priorityt" %in% sheet_names)) {
  addWorksheet(supp_wb, "FEN1_priority")
}
writeData(supp_wb,"FEN1_priority", FEN1_priority_mdd %>% 
             mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% FEN1_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
                   select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# MYRF priority variants 
if (!("MYRF_priority" %in% sheet_names)) {
  addWorksheet(supp_wb, "MYRF_priority")
}
writeData(supp_wb,"MYRF_priority", MYRF_priority_mdd %>% 
             mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% MYRF_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
                  select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# TMEM258 priority variants 
if (!("TMEM258_priority" %in% sheet_names)) {
  addWorksheet(supp_wb, "TMEM258_priority")
}
writeData(supp_wb,"TMEM258_priority", TMEM258_priority_mdd %>% 
             mutate(in_metabolite_cohort = ifelse(chr_pos_ref_alt %in% TMEM258_priority_metabol$chr_pos_ref_alt, "Yes", "No")) %>% 
                  select(-c(ends_with("_cov"), FILTER.y, CHR, POS, REF, ALT)))


# save the sheets 
saveWorkbook(supp_wb, '/Users/ellad/UniversityEdinburgh/PhD/Year_3/WGS_proj/WGS_supp_data.xlsx', overwrite=T)

```


